### ISSUE TITLE
Very slow startup after "Restore Session" on Firefox

### ISSUE THREAD

**Issue by Delgan** [uBlockOrigin/uBlock-issues#2665]
### Prerequisites

- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uAssets/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] I am running the latest version of uBlock Origin (uBO).

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

Hi.

I observed the "Restore Session" feature of Firefox takes ~20 seconds to complete on my computer when uBlock Origin is installed (while it's basically instantaneous when it's not).

I can reproduce it pretty easily:
1. Open Firefox
2. Open 5 tabs of a lightweight website
3. Type `sudo reboot` on a terminal
4. When the computer has restarted, re-launch Firefox
5. Click on "Restore Session" button proposed by Firefox
6. Observe that the website takes a very long time to load

### URL(s) where the issue occurs.

N/A

### Screenshot(s)

N/A

### Notes

I can't reproduce it with a clean new Firefox profile. But I don't understand what in my profile could cause such behavior. Does uBlock parse cookies, cache, or history at startup? I have disabled all other extensions. Also, it seems the problem does not appear if Firefox is started as usual, without session restoration.

Profiler with uBlock Origin disabled: [Firefox_2023-05-20_ublock_off.json.gz](https://github.com/uBlockOrigin/uAssets/files/11522544/Firefox_2023-05-20_ublock_off.json.gz)
Profiler with uBlock Origin enabled: [Firefox 2023-05-20 16.19 profile.json.gz](https://github.com/uBlockOrigin/uBlock-issues/files/11522557/Firefox.2023-05-20.16.19.profile.json.gz)

I tried to read logger but there seems to be no output at startup. I can't find much in the profiler either. Any clues would be welcome.

### Configuration

<details>

```yaml
uBlock Origin: 1.49.2
Firefox: 113
filterset (summary): 
 network: 93806
 cosmetic: 38741
 scriptlet: 17439
 html: 942
listset (total-discarded, last-updated): 
 default: 
  user-filters: 0-0, never
  ublock-filters: 33888-70, 3m
  ublock-badware: 7102-1, 3m
  ublock-privacy: 442-0, 3m
  ublock-abuse: 70-70, 3m
  ublock-unbreak: 2031-2, 3m
  ublock-quick-fixes: 349-0, 3m
  easylist: 65960-647, 3m
  easyprivacy: 31844-615, 3m
  urlhaus-1: 7065-1, 3m
  plowe-0: 3735-2, 3m
filterset (user): [empty]
modifiedUserSettings: 
 suspendUntilListsAreLoaded: false
modifiedHiddenSettings: [none]
supportStats: 
 allReadyAfter: 21610 ms
 maxAssetCacheWait: 398 ms
 unsuspendAfter: 20685 ms
```
</details>

**Comment by gorhill** [comment]
> Does uBlock parse cookies, cache, or history at startup?

No it does not, it just load filtering data from storage, and from what I can see in your profiling data, uBO is just idle waiting for the browser API to return the filtering data from storage, which it finally receives at 25-26s mark. Your profile has some issues and uBO is just making it more apparent because it will hold back network requests until it receives the filtering data from storage.

If you uncheck _"Suspend network activity until all filter lists are loaded"_ in _"Filter lists"_ pane, uBO won't hold back network requests until ready, but then this also mean it won't be able to filter properly at launch whichever web pages is being loaded.

So in summary you need to investigate what in your profile is causing issue, I can't do this for you, you need to seek help from Firefox community or Firefox people, uBO taking long to be ready is just a symptom of other issues in your profile. Just in case maybe uBO's storage is somehow broken, you can cause it to be re-generated by clicking _"Purge all caches"_ while holding <kbd>Shift</kbd> key.

Closing as _invalid_ since the profiling data does not show uBO being the issue.

**Comment by gwarser** [comment]
>Type `sudo reboot` on a terminal

This is your problem. You are breaking Firefox storage/settings by this. Firefox does not have code to gracefully shutdown on a signal. You need to close it manually. https://bugzilla.mozilla.org/show_bug.cgi?id=336193

**Comment by Delgan** [comment]
Thanks for both answers.

> If you uncheck "Suspend network activity until all filter lists are loaded" in "Filter lists" pane, uBO won't hold back network requests until ready, but then this also mean it won't be able to filter properly at launch whichever web pages is being loaded.

That's what I did as indicated in the configuration, and yet I observed the blocking. But I suppose uBlock Origin is waiting for something else that Firefox can't provide yet.

> This is your problem. You are breaking Firefox storage/settings by this. Firefox does not have code to gracefully shutdown on a signal. You need to close it manually. https://bugzilla.mozilla.org/show_bug.cgi?id=336193

Yes, I understand that this "Restore Session" feature is a catastrophe recovery solution. In this sense, we can expect that some APIs will not be directly available while Firefox gets itself in order, which is logically slower with a non-empty profile.

**Comment by krystian3w** [comment]
For me update Firefox now generates long delay:

```yml
supportStats:
 allReadyAfter: 35707 ms (selfie)
 maxAssetCacheWait: 506 ms
```

I tested on update from to:

- 115.1.0 ESR > 115.2.0 ESR
- 117.0.0 RC1 > 117.0.0 RC2

**Comment by ruanmed** [comment]
Hi @Delgan , sorry to "revive" this old thread of yours, but since this GitHub Issue comes up when searching for "Firefox slow startup" or similar queries I want to leave the info below here.

The issue you were experiencing was probably due to this Firefox bug (not related to uBlock Origin): 

- [Firefox first startup high disk usage, io wait, general lag 30-40sec](https://old.reddit.com/r/firefox/comments/17u92cj/firefox_first_startup_high_disk_usage_io_wait/)
- Bugzilla: [Asynchronous non-blocking temporary storage initialization tolerating broken origins](https://bugzilla.mozilla.org/show_bug.cgi?id=1671932)
- WHAT HAPPENS: Firefox starts using the disk too much, increasing disk BUSY Time even on the fastest NVMEs on market to the point the delay on startup is noticeable and hinders user experience.
- WHAT CAUSES IT: The most common cause seems to be the Firefox CACHE and however it's handled on startup that, from what I understood, causes a complete scan of all Firefox Local Websites Cache, sometimes multiple times, unnecessarily before any webpage is loaded. The other cause seems to be if you have many FONTs installed, and Firefox might hang reading all your fonts...

A potential TEMPORARY solution currently is to erase your browser CACHE. (and btw, that why it usually does not reproduce in New Profiles, since it's way more common for users to NOT delete cache over the months/years than to install thousands of fonts on their systems).

The bug is over 4 years old, however there has been some advancements recently on fixing the issue (17 days ago there was activity on patchsets submitted to related blocking issues that this issue depends on), it seems :pray:.
